<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="Joyce Hsiao" />


<title>Compute variance explained under different K</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-4.5.0/css/font-awesome.min.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}

.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->




<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
  padding-left: 25px;
  text-indent: 0;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">nasalmicrobiome</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jhsiao999/nasalmicrobiome">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Compute variance explained under different K</h1>
<h4 class="author"><em>Joyce Hsiao</em></h4>

</div>


<!-- The file analysis/chunks.R contains chunks that define default settings
shared across the workflowr files. -->
<!-- Update knitr chunk options -->
<!-- Insert the date the file was last updated -->
<pre><code>Warning in as.POSIXlt.POSIXct(Sys.time()): unknown timezone &#39;zone/tz/2018c.
1.0/zoneinfo/America/Chicago&#39;</code></pre>
<p><strong>Last updated:</strong> 2018-06-07</p>
<!-- Insert the code version (Git commit SHA1) if Git repository exists and R
 package git2r is installed -->
<p><strong>Code version:</strong> 47a3967</p>
<div id="background-and-summary" class="section level2">
<h2>Background and summary</h2>
<p>We compare fitting under differet k and computed for every feature, sample variance in abundance that is explained by sample cluster membership.</p>
<p>Results showed that k=2 gives the largest variance explained (median of sample variances is higher than the other k).</p>
<hr />
</div>
<div id="analysis" class="section level2">
<h2>Analysis</h2>
<pre class="r"><code>library(ggplot2)
library(metagenomeSeq)</code></pre>
<p>Load data</p>
<pre class="r"><code>MRobj &lt;- readRDS(&quot;../data/nasal_filtered_normed_batchcorrected.rds&quot;)
counts &lt;- MRcounts(MRobj,norm=FALSE,log=FALSE)</code></pre>
<p>Load previously computed results.</p>
<pre class="r"><code>clust_2 &lt;- readRDS(&quot;../output/gom-k2-best-seed-67.rds&quot;)
clust_3 &lt;- readRDS(&quot;../output/gom-k3-best-seed-67.rds&quot;)
clust_4 &lt;- readRDS(&quot;../output/gom-k4-best-seed-59.rds&quot;)
clust_5 &lt;- readRDS(&quot;../output/gom-k5-best-seed-307.rds&quot;)
fits &lt;- c(clust_2, clust_3, clust_4, clust_5)
names(fits) &lt;- c(&quot;clust_2&quot;, &quot;clust_3&quot;, &quot;clust_4&quot;, &quot;clust_5&quot;)</code></pre>
<p>Extract sampel cluster membership.</p>
<pre class="r"><code>membership &lt;- lapply(fits, function(xx) {
     apply(xx$omega, 1, which.max)
  })
names(membership) &lt;- c(&quot;clust_2&quot;, &quot;clust_3&quot;, &quot;clust_4&quot;, &quot;clust_5&quot;)</code></pre>
<p>Compute variance explained.</p>
<pre class="r"><code>varprop_species &lt;- lapply(1:length(membership), function(index) {
  which_sample &lt;- which(colnames(counts) %in% names(membership[[index]]))
  counts_which &lt;- counts[, which_sample]
  var &lt;- sapply(1:nrow(counts_which), function(index2) {
            res &lt;- lm(log2(counts_which[index2,]+2) ~ membership[[index]])
            sumOfsquares &lt;- anova(res)[[2]]
            sumOfsquares[1]/sum(sumOfsquares) })
  return(var)
})
names(varprop_species) &lt;- names(membership)
saveRDS(varprop_species, file = &quot;../data/varprop.rds&quot;)</code></pre>
<p>Plot variance explained</p>
<pre class="r"><code>varprop &lt;- readRDS(&quot;../data/varprop.rds&quot;)
varprop_long &lt;- data.frame(
  prop = do.call(c, varprop),
  cluster = factor(rep(2:5, each = nrow(varprop))) )
varprop_long$cluster &lt;- as.numeric(as.character(varprop_long$cluster))</code></pre>
<p>K = 2 to 5</p>
<pre class="r"><code>ggplot(varprop_long, aes(x = factor(cluster), y = prop,
                            col = factor(cluster)), size = .2) +
  geom_violin(trim = FALSE, size=.5) +
  geom_boxplot(width=0.1, size = .6) +
  labs(x = &quot;Number of clusters&quot;, y = &quot;Proportion variance explained&quot;)</code></pre>
<p><img src="figure/cluster-variance-explained.Rmd/unnamed-chunk-7-1.png" width="672" style="display: block; margin: auto;" /></p>
<hr />
</div>
<div id="compare-variances-explained" class="section level2">
<h2>Compare variances explained</h2>
<p>Sig difference between K=2 vs K=4,5, but no sig. difference between K=2 and K=3.</p>
<pre class="r"><code>wilcox.test(varprop$clust_2, varprop$clust_3)</code></pre>
<pre><code>
    Wilcoxon rank sum test with continuity correction

data:  varprop$clust_2 and varprop$clust_3
W = 185720, p-value = 6.513e-10
alternative hypothesis: true location shift is not equal to 0</code></pre>
<pre class="r"><code>wilcox.test(varprop$clust_2, varprop$clust_4)</code></pre>
<pre><code>
    Wilcoxon rank sum test with continuity correction

data:  varprop$clust_2 and varprop$clust_4
W = 183850, p-value = 5.653e-09
alternative hypothesis: true location shift is not equal to 0</code></pre>
<pre class="r"><code>wilcox.test(varprop$clust_2, varprop$clust_5)</code></pre>
<pre><code>
    Wilcoxon rank sum test with continuity correction

data:  varprop$clust_2 and varprop$clust_5
W = 166400, p-value = 0.01104
alternative hypothesis: true location shift is not equal to 0</code></pre>
<p>Consider within-sample diversity difference for when K=3.</p>
<pre class="r"><code>library(data.table)
shannon_values &lt;- fread(&quot;~/Dropbox/current-projects/nasalMicrobiomeManuscript/draft-results/shannon_values.tsv&quot;)

all.equal(shannon_values$V1, rownames(fits$clust_3$omega))</code></pre>
<pre><code>[1] TRUE</code></pre>
<pre class="r"><code>shannon_values$GOM &lt;- sapply(1:nrow(fits$clust_3$omega), function(i) {
  which.max(fits$clust_3$omega[i,]) })

boxplot(shannon_values$V2~ shannon_values$GOM)</code></pre>
<p><img src="figure/cluster-variance-explained.Rmd/unnamed-chunk-9-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>levels(shannon_values$GOM) &lt;- c(&quot;Group 1&quot;, &quot;Group 2&quot;, &quot;Group 3&quot;)
library(ggplot2)
ggplot(shannon_values, aes(x=GOM, y = V2, fill = factor(GOM))) +
  geom_violin(col = &quot;gray50&quot;) +
  scale_fill_manual(values = c(brewer.pal(12,&quot;Set3&quot;)[1],
                                 brewer.pal(12,&quot;Set3&quot;)[2],
                                 brewer.pal(12,&quot;Set3&quot;)[3])) +
  geom_boxplot(width=.25) + ylab(&quot;Shannon diversity&quot;) +
  scale_x_discrete(name = &quot;Cluster membership&quot;,
                   labels = c(&quot;Group 1&quot;, &quot;Group 2&quot;, &quot;Group 3&quot;)) +
  guides(fill=guide_legend(title=&quot;&quot;)) +
  ggtitle(&quot;Alpha diversity of K=3&quot;)</code></pre>
<p><img src="figure/cluster-variance-explained.Rmd/unnamed-chunk-9-2.png" width="672" style="display: block; margin: auto;" /></p>
<p>Consider within-sample diversity difference for when K=4.</p>
<pre class="r"><code>library(data.table)
shannon_values &lt;- fread(&quot;~/Dropbox/current-projects/nasalMicrobiomeManuscript/draft-results/shannon_values.tsv&quot;)

all.equal(shannon_values$V1, rownames(fits$clust_4$omega))</code></pre>
<pre><code>[1] TRUE</code></pre>
<pre class="r"><code>shannon_values$GOM &lt;- sapply(1:nrow(fits$clust_4$omega), function(i) {
  which.max(fits$clust_4$omega[i,]) })

levels(shannon_values$GOM) &lt;- c(&quot;Group 1&quot;, &quot;Group 2&quot;, &quot;Group 3&quot;, &quot;Group 4&quot;)
library(ggplot2)
ggplot(shannon_values, aes(x=GOM, y = V2, fill = factor(GOM))) +
  geom_violin(col = &quot;gray50&quot;) +
  scale_fill_manual(values = c(brewer.pal(12,&quot;Set3&quot;)[1],
                                 brewer.pal(12,&quot;Set3&quot;)[2],
                                 brewer.pal(12,&quot;Set3&quot;)[3],
                                 brewer.pal(12,&quot;Set3&quot;)[4])) +
  geom_boxplot(width=.25) + ylab(&quot;Shannon diversity&quot;) +
  scale_x_discrete(name = &quot;Cluster membership&quot;,
                   labels = c(&quot;Group 1&quot;, &quot;Group 2&quot;, &quot;Group 3&quot;, &quot;Group 4&quot;)) +
  guides(fill=guide_legend(title=&quot;&quot;)) +
  ggtitle(&quot;Alpha diversity of K=4&quot;)</code></pre>
<p><img src="figure/cluster-variance-explained.Rmd/unnamed-chunk-10-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Consider within-sample diversity difference for when K=5.</p>
<pre class="r"><code>library(data.table)
shannon_values &lt;- fread(&quot;~/Dropbox/current-projects/nasalMicrobiomeManuscript/draft-results/shannon_values.tsv&quot;)

all.equal(shannon_values$V1, rownames(fits$clust_5$omega))</code></pre>
<pre><code>[1] TRUE</code></pre>
<pre class="r"><code>shannon_values$GOM &lt;- sapply(1:nrow(fits$clust_5$omega), function(i) {
  which.max(fits$clust_5$omega[i,]) })

boxplot(shannon_values$V2~ shannon_values$GOM)</code></pre>
<p><img src="figure/cluster-variance-explained.Rmd/unnamed-chunk-11-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>levels(shannon_values$GOM) &lt;- c(&quot;Group 1&quot;, &quot;Group 2&quot;, &quot;Group 3&quot;, &quot;Group 4&quot;, &quot;Group 5&quot;)
library(ggplot2)
ggplot(shannon_values, aes(x=GOM, y = V2, fill = factor(GOM))) +
  geom_violin(col = &quot;gray50&quot;) +
  scale_fill_manual(values = c(brewer.pal(12,&quot;Set3&quot;)[1],
                                 brewer.pal(12,&quot;Set3&quot;)[2],
                                 brewer.pal(12,&quot;Set3&quot;)[3],
                                 brewer.pal(12,&quot;Set3&quot;)[4],
                                 brewer.pal(12,&quot;Set3&quot;)[5])) +
  geom_boxplot(width=.25) + ylab(&quot;Shannon diversity&quot;) +
  scale_x_discrete(name = &quot;Cluster membership&quot;,
                   labels = c(&quot;Group 1&quot;, &quot;Group 2&quot;, &quot;Group 3&quot;, &quot;Group 4&quot;, &quot;Group 5&quot;)) +
  guides(fill=guide_legend(title=&quot;&quot;)) +
  ggtitle(&quot;Alpha diversity of K=5&quot;)</code></pre>
<p><img src="figure/cluster-variance-explained.Rmd/unnamed-chunk-11-2.png" width="672" style="display: block; margin: auto;" /></p>
<hr />
</div>
<div id="consider-when-k2-the-taxa-that-has-high-prob-of-variance-explained" class="section level2">
<h2>Consider when K=2, the taxa that has high prob of variance explained</h2>
<pre class="r"><code>rownames(counts)[order(varprop$clust_2, decreasing = F)[1:10]]</code></pre>
<pre><code> [1] &quot;k__Bacteria;p__Proteobacteria;c__Betaproteobacteria;o__Hydrogenophilales;f__Hydrogenophilaceae;g__Thiobacillus;s__&quot;
 [2] &quot;k__Bacteria;p__Firmicutes;c__Clostridia;o__Clostridiales;f__Veillonellaceae;g__Veillonella;s__&quot;                    
 [3] &quot;k__Bacteria;p__Actinobacteria;c__Actinobacteria;o__Actinomycetales;f__Nocardioidaceae;g__Friedmanniella;s__&quot;       
 [4] &quot;k__Bacteria;p__Actinobacteria;c__Actinobacteria;o__Actinomycetales;f__Geodermatophilaceae;g__Modestobacter;s__&quot;    
 [5] &quot;k__Bacteria;p__Bacteroidetes;c__[Saprospirae];o__[Saprospirales];f__Chitinophagaceae;g__Sediminibacterium;s__&quot;     
 [6] &quot;k__Bacteria;p__Actinobacteria;c__Actinobacteria;o__Actinomycetales;f__Cellulomonadaceae;g__Cellulomonas;s__&quot;       
 [7] &quot;s__subflava&quot;                                                                                                       
 [8] &quot;k__Bacteria;p__Firmicutes;c__Bacilli;o__Bacillales;f__Bacillaceae;g__Geobacillus;s__&quot;                              
 [9] &quot;k__Bacteria;p__Proteobacteria;c__Gammaproteobacteria;o__Enterobacteriales;f__Enterobacteriaceae;g__Klebsiella;s__&quot; 
[10] &quot;k__Bacteria;p__Proteobacteria;c__Betaproteobacteria;o__Burkholderiales;f__Comamonadaceae;g__Comamonas;s__&quot;         </code></pre>
<pre class="r"><code>library(data.table)
de_results &lt;- fread(&quot;~/Dropbox/current-projects/nasalMicrobiomeManuscript/draft-results/tbl.tsv&quot;)

all.equal(de_results$V1[match(rownames(counts), de_results$V1)], rownames(counts))</code></pre>
<pre><code>[1] TRUE</code></pre>
<pre class="r"><code>de_results &lt;- de_results[match(rownames(counts), de_results$V1),]

plot(varprop$clust_2,-log10(de_results$adjPvalue),
     xlab = &quot;Proportion of variance explained of K=2&quot;,
     ylab=&quot;-log10(adjusted P-value)&quot;,
     pch=16, cex=.7,
     col=(de_results$adjPvalue&lt;.01)+1)
abline(h=2,lty=2,col=&quot;black&quot;)</code></pre>
<p><img src="figure/cluster-variance-explained.Rmd/unnamed-chunk-12-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>de_results$varprop_clust2 &lt;- varprop$clust_2</code></pre>
<p>Append the variance proportion explained to the FDR results.</p>
<pre class="r"><code>de_results &lt;- de_results[order(de_results$adjPvalues),]
write.table(de_results,
            &quot;~/Dropbox/current-projects/nasalMicrobiomeManuscript/draft-results/tbl_updated.tsv&quot;,
            quote=F, col.names = NA, sep=&quot;\t&quot;)

#tmp &lt;- fread(&quot;~/Dropbox/current-projects/nasalMicrobiomeManuscript/draft-results/tbl_updated.tsv&quot;)</code></pre>
<hr />
</div>
<div id="session-information" class="section level2">
<h2>Session information</h2>
<pre><code>R version 3.4.1 (2017-06-30)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS High Sierra 10.13

Matrix products: default
BLAS: /Library/Frameworks/R.framework/Versions/3.4/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.4/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] parallel  stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] data.table_1.11.0    metagenomeSeq_1.21.1 RColorBrewer_1.1-2  
 [4] glmnet_2.0-16        foreach_1.4.4        Matrix_1.2-14       
 [7] limma_3.34.9         Biobase_2.38.0       BiocGenerics_0.24.0 
[10] ggplot2_2.2.1       

loaded via a namespace (and not attached):
 [1] Rcpp_0.12.16       pillar_1.2.2       compiler_3.4.1    
 [4] git2r_0.21.0       plyr_1.8.4         bitops_1.0-6      
 [7] iterators_1.0.9    tools_3.4.1        digest_0.6.15     
[10] evaluate_0.10.1    tibble_1.4.2       gtable_0.2.0      
[13] lattice_0.20-35    rlang_0.2.0        yaml_2.1.19       
[16] stringr_1.3.0      knitr_1.20         caTools_1.17.1    
[19] gtools_3.5.0       rprojroot_1.3-2    grid_3.4.1        
[22] rmarkdown_1.9      gdata_2.18.0       magrittr_1.5      
[25] backports_1.1.2    scales_0.5.0       gplots_3.0.1      
[28] codetools_0.2-15   htmltools_0.3.6    matrixStats_0.53.1
[31] colorspace_1.3-2   labeling_0.3       KernSmooth_2.23-15
[34] stringi_1.2.2      lazyeval_0.2.1     munsell_0.4.3     </code></pre>
</div>

<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>

<hr>
<p>
    This <a href="http://rmarkdown.rstudio.com">R Markdown</a> site was created with <a href="https://github.com/jdblischak/workflowr">workflowr</a>
</p>
<hr>

<!-- To enable disqus, uncomment the section below and provide your disqus_shortname -->

<!-- disqus
  <div id="disqus_thread"></div>
    <script type="text/javascript">
        /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
        var disqus_shortname = 'rmarkdown'; // required: replace example with your forum shortname

        /* * * DON'T EDIT BELOW THIS LINE * * */
        (function() {
            var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
            dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
            (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
        })();
    </script>
    <noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
    <a href="http://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>
-->


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
