<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="JNP" />


<title>Differential abundance analysis at higher taxa levels</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.0.13/css/fa-svg-with-js.css" rel="stylesheet" />
<script src="site_libs/font-awesome-5.0.13/js/fontawesome-all.min.js"></script>
<script src="site_libs/font-awesome-5.0.13/js/fa-v4-shims.min.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}

.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->




<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
  padding-left: 25px;
  text-indent: 0;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">nasalmicrobiome</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jhsiao999/nasalmicrobiome">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Differential abundance analysis at higher taxa levels</h1>
<h4 class="author"><em>JNP</em></h4>

</div>


<!-- The file analysis/chunks.R contains chunks that define default settings
shared across the workflowr files. -->
<!-- Update knitr chunk options -->
<!-- Insert the date the file was last updated -->
<!-- Insert the code version (Git commit SHA1) if Git repository exists and R
 package git2r is installed -->
<hr />
<div id="loading-data-and-packages" class="section level2">
<h2>Loading data and packages</h2>
<pre class="r"><code>library(knitr)
library(kableExtra)
library(dplyr)
library(metagenomeSeq)
library(CountClust)</code></pre>
<p>Read in filtered, corrected, etc data.</p>
<pre class="r"><code>load(&quot;../data/nasal_GOM.rdata&quot;)
colnames(fData(MRobj))</code></pre>
<hr />
</div>
<div id="aggregatebytaxonomy-code-edits" class="section level2">
<h2>aggregateByTaxonomy code edits</h2>
<pre class="r"><code>aggregateByTaxonomy &lt;- function (obj, lvl, alternate = FALSE, norm = FALSE, log = FALSE, 
    aggfun = colSums, sl = 1000, featureOrder = NULL, returnFullHierarchy = TRUE, 
    out = &quot;MRexperiment&quot;) 
{
    if (class(obj) == &quot;MRexperiment&quot;) {
        mat = MRcounts(obj, norm = norm, log = log, sl = sl)
        if (length(lvl) == 1) {
            # added code
            include_features &lt;- fData(obj)[match(rownames(mat),fData(obj)$Species),]
            levels=as.character(include_features[, lvl])
            #            levels = as.character(fData(obj)[, lvl])
        } else { levels = as.character(lvl) }
    }
    else {
        mat = obj
        levels = as.character(lvl)
        if (length(levels) != nrow(mat)) 
            stop(&quot;If input is a count matrix, lvl must be a vector of length = nrow(count matrix)&quot;)
    }
    if (!(out %in% c(&quot;MRexperiment&quot;, &quot;matrix&quot;))) {
        stop(&quot;The variable out must either be &#39;MRexperiment&#39; or &#39;matrix&#39;&quot;)
    }
    nafeatures = is.na(levels)
    if (length(nafeatures) &gt; 0) {
        if (alternate == FALSE) {
            levels[nafeatures] = &quot;no_match&quot;
        }
        else {
            levels[nafeatures] = paste(&quot;OTU_&quot;, rownames(obj)[nafeatures], 
                sep = &quot;&quot;)
        }
    }
    grps = split(seq_along(levels), levels)
    newMat = array(NA, dim = c(length(grps), ncol(obj)))
    for (i in seq_along(grps)) {
        newMat[i, ] = aggfun(mat[grps[[i]], , drop = FALSE])
    }
    rownames(newMat) = names(grps)
    colnames(newMat) = colnames(obj)
    if (out == &quot;matrix&quot;) 
        return(newMat)
    if (out == &quot;MRexperiment&quot;) {
        if (returnFullHierarchy) {
            if (is.null(featureOrder)) {
                featureOrder &lt;- colnames(fData(obj))
            }
            taxa = featureData(obj)[match(names(grps), fData(obj)[, 
                lvl]), featureOrder[1:which(featureOrder == lvl)]]
            featureNames(taxa) = names(grps)
        }
        else {
            taxa = data.frame(names(grps))
            colnames(taxa) = &quot;Taxa&quot;
            rownames(taxa) = names(grps)
            taxa = as(taxa, &quot;AnnotatedDataFrame&quot;)
        }
        if (class(obj) == &quot;MRexperiment&quot;) {
            pd = phenoData(obj)
            newObj = newMRexperiment(newMat, featureData = taxa, 
                phenoData = pd)
        }
        else {
            newObj = newMRexperiment(newMat, featureData = taxa)
        }
        return(newObj)
    }
}</code></pre>
</div>
<div id="analysis-at-genus-level" class="section level2">
<h2>Analysis at genus level</h2>
<pre class="r"><code>cl = pData(MRobj)$GOM
mod = model.matrix(~cl)

obj &lt;- aggregateByTaxonomy(MRobj, lvl=&quot;Genus&quot;, aggfun=colSums)

# export aggregated count table
# write.xlsx(MRcounts(obj), file=&quot;../data/counts_genus.xlsx&quot;,
#            col.names=TRUE, row.names=TRUE, append=FALSE, showNA=TRUE, password=NULL)
obj@expSummary$expSummary$normFactors &lt;- normFactors(MRobj)
fit &lt;- fitFeatureModel(obj, mod=mod)
MRfulltable(fit,number = nrow(obj),file=&quot;../data/tbl_genus.tsv&quot;)
dim(obj)</code></pre>
<p>â€“</p>
</div>
<div id="analysis-at-family-level" class="section level2">
<h2>Analysis at family level</h2>
<pre class="r"><code>cl = pData(MRobj)$GOM
mod = model.matrix(~cl)

obj &lt;- aggregateByTaxonomy(MRobj, lvl=&quot;Family&quot;, aggfun=colSums)
# # export aggregated count table
# write.xlsx(MRcounts(obj), file=&quot;../data/counts_family.xlsx&quot;,
#            col.names=TRUE, row.names=TRUE, append=FALSE, showNA=TRUE, password=NULL)
obj@expSummary$expSummary$normFactors &lt;- normFactors(MRobj)
fit &lt;- fitFeatureModel(obj, mod=mod)
MRfulltable(fit,number = nrow(obj),file=&quot;../data/tbl_family.tsv&quot;)
dim(obj)</code></pre>
<hr />
</div>
<div id="analysis-at-order-level" class="section level2">
<h2>Analysis at order level</h2>
<pre class="r"><code>cl = pData(MRobj)$GOM
mod = model.matrix(~cl)

obj &lt;- aggregateByTaxonomy(MRobj, lvl=&quot;Order&quot;, aggfun=colSums)
# # export aggregated count table
# write.xlsx(MRcounts(obj), file=&quot;../data/counts_order.xlsx&quot;,
#            col.names=TRUE, row.names=TRUE, append=FALSE, showNA=TRUE, password=NULL)

obj@expSummary$expSummary$normFactors &lt;- normFactors(MRobj)
fit &lt;- fitFeatureModel(obj, mod=mod)
MRfulltable(fit,number = nrow(obj),file=&quot;../data/tbl_order.tsv&quot;)
dim(obj)</code></pre>
</div>
<div id="analysis-at-class-level" class="section level2">
<h2>Analysis at class level</h2>
<pre class="r"><code>cl = pData(MRobj)$GOM
mod = model.matrix(~cl)

obj &lt;- aggregateByTaxonomy(MRobj, lvl=&quot;Class&quot;, aggfun=colSums)
# # export aggregated count table
# write.xlsx(MRcounts(obj), file=&quot;../data/counts_class.xlsx&quot;,
#            col.names=TRUE, row.names=TRUE, append=FALSE, showNA=TRUE, password=NULL)

obj@expSummary$expSummary$normFactors &lt;- normFactors(MRobj)
fit &lt;- fitFeatureModel(obj, mod=mod)
MRfulltable(fit,number = nrow(obj),file=&quot;../data/tbl_class.tsv&quot;)
dim(obj)</code></pre>
<hr />
</div>
<div id="analysis-at-phylum-level" class="section level2">
<h2>Analysis at phylum level</h2>
<pre class="r"><code>cl = pData(MRobj)$GOM
mod = model.matrix(~cl)

obj &lt;- aggregateByTaxonomy(MRobj, lvl=&quot;Phylum&quot;, aggfun=colSums)
obj@expSummary$expSummary$normFactors &lt;- normFactors(MRobj)
fit &lt;- fitFeatureModel(obj, mod=mod)
MRfulltable(fit,number = nrow(obj),file=&quot;../data/tbl_phylum.tsv&quot;)
dim(obj)</code></pre>
<hr />
</div>
<div id="session-info" class="section level2">
<h2>Session Info</h2>
<pre><code>## R version 3.4.1 (2017-06-30)
## Platform: x86_64-apple-darwin15.6.0 (64-bit)
## Running under: macOS High Sierra 10.13
## 
## Matrix products: default
## BLAS: /Library/Frameworks/R.framework/Versions/3.4/Resources/lib/libRblas.0.dylib
## LAPACK: /Library/Frameworks/R.framework/Versions/3.4/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
## [1] parallel  stats     graphics  grDevices utils     datasets  methods  
## [8] base     
## 
## other attached packages:
##  [1] CountClust_1.4.1     ggplot2_3.0.0        metagenomeSeq_1.21.1
##  [4] RColorBrewer_1.1-2   glmnet_2.0-16        foreach_1.4.4       
##  [7] Matrix_1.2-14        limma_3.34.9         Biobase_2.38.0      
## [10] BiocGenerics_0.24.0  dplyr_0.7.6          kableExtra_0.9.0    
## [13] knitr_1.20          
## 
## loaded via a namespace (and not attached):
##  [1] httr_1.3.1         viridisLite_0.3.0  gtools_3.8.1      
##  [4] assertthat_0.2.0   stats4_3.4.1       yaml_2.2.0        
##  [7] slam_0.1-43        pillar_1.3.0       backports_1.1.2   
## [10] lattice_0.20-35    glue_1.3.0         digest_0.6.17     
## [13] rvest_0.3.2        colorspace_1.3-2   picante_1.7       
## [16] cowplot_0.9.3      htmltools_0.3.6    plyr_1.8.4        
## [19] pkgconfig_2.0.2    purrr_0.2.5        scales_1.0.0      
## [22] gdata_2.18.0       tibble_1.4.2       mgcv_1.8-24       
## [25] withr_2.1.2        nnet_7.3-12        lazyeval_0.2.1    
## [28] magrittr_1.5       crayon_1.3.4       evaluate_0.12     
## [31] MASS_7.3-50        nlme_3.1-137       gplots_3.0.1      
## [34] xml2_1.2.0         vegan_2.5-2        tools_3.4.1       
## [37] hms_0.4.2          matrixStats_0.54.0 stringr_1.3.1     
## [40] munsell_0.5.0      cluster_2.0.7-1    bindrcpp_0.2.2    
## [43] compiler_3.4.1     caTools_1.17.1.1   rlang_0.2.2       
## [46] grid_3.4.1         iterators_1.0.10   rstudioapi_0.8    
## [49] bitops_1.0-6       rmarkdown_1.10     boot_1.3-20       
## [52] gtable_0.2.0       codetools_0.2-15   flexmix_2.3-14    
## [55] reshape2_1.4.3     R6_2.3.0           bindr_0.1.1       
## [58] rprojroot_1.3-2    maptpx_1.9-3       permute_0.9-4     
## [61] KernSmooth_2.23-15 readr_1.1.1        ape_5.2           
## [64] modeltools_0.2-22  stringi_1.2.4      SQUAREM_2017.10-1 
## [67] Rcpp_0.12.19       tidyselect_0.2.4</code></pre>
</div>

<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>

<hr>
<p>
    This <a href="http://rmarkdown.rstudio.com">R Markdown</a> site was created with <a href="https://github.com/jdblischak/workflowr">workflowr</a>
</p>
<hr>

<!-- To enable disqus, uncomment the section below and provide your disqus_shortname -->

<!-- disqus
  <div id="disqus_thread"></div>
    <script type="text/javascript">
        /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
        var disqus_shortname = 'rmarkdown'; // required: replace example with your forum shortname

        /* * * DON'T EDIT BELOW THIS LINE * * */
        (function() {
            var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
            dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
            (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
        })();
    </script>
    <noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
    <a href="http://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>
-->


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
