<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="Joyce Hsiao" />


<title>Investigate clusering probability in relation to PCA results</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-4.5.0/css/font-awesome.min.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}

.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->




<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
  padding-left: 25px;
  text-indent: 0;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">nasalmicrobiome</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jhsiao999/nasalmicrobiome">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Investigate clusering probability in relation to PCA results</h1>
<h4 class="author"><em>Joyce Hsiao</em></h4>

</div>


<!-- The file analysis/chunks.R contains chunks that define default settings
shared across the workflowr files. -->
<!-- Update knitr chunk options -->
<!-- Insert the date the file was last updated -->
<pre><code>Warning in as.POSIXlt.POSIXct(Sys.time()): unknown timezone &#39;zone/tz/2018c.
1.0/zoneinfo/America/Chicago&#39;</code></pre>
<p><strong>Last updated:</strong> 2018-06-07</p>
<!-- Insert the code version (Git commit SHA1) if Git repository exists and R
 package git2r is installed -->
<p><strong>Code version:</strong> 867e392</p>
<hr />
<div id="background-and-summary" class="section level2">
<h2>Background and summary</h2>
<p>Examine the GOM membership probability of the samples that are outliers on the PC1 vs PC2 plot.</p>
<hr />
</div>
<div id="analysis" class="section level2">
<h2>Analysis</h2>
<pre class="r"><code>library(ggplot2)
library(metagenomeSeq)
library(matrixStats)</code></pre>
<p>Load PCA</p>
<pre class="r"><code>obj &lt;- readRDS(&quot;../data/nasal_GOM.rds&quot;)</code></pre>
<p>Load previously computed results for k=2.</p>
<pre class="r"><code>fits &lt;- readRDS(file = &quot;../data/count-clust.rds&quot;)</code></pre>
<p>Get PCA results</p>
<pre class="r"><code>pcaRES &lt;- function(obj,tran=TRUE,comp=1:2,norm=TRUE,log=TRUE,usePCA=TRUE,useDist=FALSE,distfun=stats::dist,dist.method=&quot;euclidian&quot;,n=NULL,...){
  mat = returnAppropriateObj(obj,norm,log)
  if(useDist==FALSE &amp; usePCA==FALSE) stop(&quot;Classical MDS requires distances&quot;)
  if(is.null(n)) n = min(nrow(mat),1000)

  otusToKeep &lt;- which(rowSums(mat)&gt;0)
  otuVars&lt;-rowSds(mat[otusToKeep,])
  otuIndices&lt;-otusToKeep[order(otuVars,decreasing=TRUE)[seq_len(n)]]
  mat &lt;- mat[otuIndices,]

  if(tran==TRUE){
    mat = t(mat)
  }
  if(useDist==TRUE){
    d &lt;- distfun(mat,method=dist.method)
  } else{ d = mat }

  if(usePCA==FALSE){
    ord = cmdscale(d,k = max(comp))
    xl = paste(&quot;MDS component:&quot;,comp[1])
    yl = paste(&quot;MDS component:&quot;,comp[2])
  } else{
    pcaRes &lt;- prcomp(d)
    ord &lt;- pcaRes$x
    vars &lt;- pcaRes$sdev^2
    vars &lt;- round(vars/sum(vars),5)*100
    xl &lt;- sprintf(&quot;PCA %s: %.2f%% variance&quot;,colnames(ord)[comp[1]], vars[comp[1]])
    yl &lt;- sprintf(&quot;PCA %s: %.2f%% variance&quot;,colnames(ord)[comp[2]], vars[comp[2]])
  }
  return(pcaRes)
}

obj &lt;- obj[,order(pData(obj)$GOM)]

pcares &lt;- pcaRES(obj,pch=21,bg=pData(obj)$GOM,main=&quot;&quot;)</code></pre>
<p>save to output folder.</p>
<pre class="r"><code>saveRDS(pcares, &quot;../output/explore-cluster-probability.Rmd/pcares.rds&quot;)</code></pre>
</div>
<div id="results" class="section level2">
<h2>Results</h2>
<p>7 outliers on PCA plot. All of them below to the second cluster.</p>
<pre class="r"><code>all.equal(rownames(pcares$x), colnames(obj))</code></pre>
<pre><code>[1] TRUE</code></pre>
<pre class="r"><code>pdata &lt;- pData(obj)
pdata$PC1 &lt;- pcares$x[,1]
pdata$PC2 &lt;- pcares$x[,2]
pdata$pc_outlier &lt;- pdata$PC1 &gt; 20 &amp; pdata$PC2 &lt; -20

all.equal(rownames(pdata), rownames(fits[[1]]$omega))</code></pre>
<pre><code>[1] &quot;186 string mismatches&quot;</code></pre>
<pre class="r"><code>omega &lt;- fits[[1]]$omega
omega &lt;- omega[match(rownames(pdata),rownames(omega)),]
all.equal(rownames(pdata), rownames(omega))</code></pre>
<pre><code>[1] TRUE</code></pre>
<pre class="r"><code>pdata$gom_prob_group1 &lt;- omega[,1]
pdata$gom_prob_group2 &lt;- omega[,2]</code></pre>
<ul>
<li>7 PCA outliers, all belong to cluster 2, their cluster probability are not borderline.</li>
</ul>
<pre class="r"><code>table(pdata$pc_outlier)</code></pre>
<pre><code>
FALSE  TRUE 
  190     7 </code></pre>
<pre class="r"><code>table(pdata$pc_outlier, pdata$GOM)</code></pre>
<pre><code>       
          1   2
  FALSE 167  23
  TRUE    0   7</code></pre>
<pre class="r"><code>cbind(pdata$gom_prob_group1[which(pdata$pc_outlier==1)], pdata$gom_prob_group2[which(pdata$pc_outlier==1)])</code></pre>
<pre><code>             [,1]      [,2]
[1,] 0.0273438612 0.9726561
[2,] 0.0001033176 0.9998967
[3,] 0.2372200486 0.7627800
[4,] 0.3211409734 0.6788590
[5,] 0.3253818129 0.6746182
[6,] 0.0785997584 0.9214002
[7,] 0.0233961031 0.9766039</code></pre>
<ul>
<li>Are their borderline probablities at all?</li>
</ul>
<pre class="r"><code>summary(pdata$gom_prob_group2[pdata$GOM==2 &amp; pdata$pc_outlier==0])</code></pre>
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
 0.5175  0.6140  0.6835  0.7184  0.8221  0.9998 </code></pre>
<pre class="r"><code>summary(pdata$gom_prob_group2[pdata$GOM==2 &amp; pdata$pc_outlier==1])</code></pre>
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
 0.6746  0.7208  0.9214  0.8553  0.9746  0.9999 </code></pre>
<p>The non-outliers in Group 2 are more similar to Group 1.</p>
<pre class="r"><code>library(CountClust)
newgroup &lt;- pdata$GOM
newgroup[which(pdata$pc_outlier==1 &amp; pdata$GOM==&quot;2&quot;)] &lt;- &quot;2.outlier&quot;
annotation &lt;- data.frame(
  sample_id = paste0(&quot;X&quot;, c(1:NROW(omega))),
  tissue_label = factor(newgroup,
                        levels = c(&quot;1&quot;, &quot;2&quot;, &quot;2.outlier&quot;) ) ) 
rownames(omega) &lt;- annotation$sample_id;

StructureGGplot(omega = omega,
                annotation = annotation,
                palette = RColorBrewer::brewer.pal(3, &quot;Accent&quot;),
                yaxis_label = &quot;Group&quot;,
                order_sample = TRUE,
                axis_tick = list(axis_ticks_length = .1,
                                 axis_ticks_lwd_y = .1,
                                 axis_ticks_lwd_x = .1,
                                 axis_label_size = 7,
                                 axis_label_face = &quot;bold&quot;))</code></pre>
<p><img src="figure/explore-cluster-probability-.Rmd/unnamed-chunk-9-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Save outlier info.</p>
<pre class="r"><code>write.table(data.frame(sample_ID=rownames(pdata),
                       gom_prob_group1=pdata$gom_prob_group1,
                       gom_prob_group2=pdata$gom_prob_group2,
                       newgroup)[order(newgroup, decreasing = T),],
            col.names = T, row.names = F, quote = F,
            file = &quot;../output/explore-cluster-probability.Rmd/outlier.list.txt&quot;)</code></pre>
<hr />
</div>
<div id="session-information" class="section level2">
<h2>Session information</h2>
<pre><code>R version 3.4.1 (2017-06-30)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS High Sierra 10.13

Matrix products: default
BLAS: /Library/Frameworks/R.framework/Versions/3.4/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.4/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] parallel  stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] CountClust_1.4.1     matrixStats_0.53.1   metagenomeSeq_1.21.1
 [4] RColorBrewer_1.1-2   glmnet_2.0-16        foreach_1.4.4       
 [7] Matrix_1.2-14        limma_3.34.9         Biobase_2.38.0      
[10] BiocGenerics_0.24.0  ggplot2_2.2.1       

loaded via a namespace (and not attached):
 [1] gtools_3.5.0       modeltools_0.2-21  slam_0.1-43       
 [4] reshape2_1.4.3     lattice_0.20-35    colorspace_1.3-2  
 [7] htmltools_0.3.6    stats4_3.4.1       mgcv_1.8-23       
[10] yaml_2.1.19        rlang_0.2.0        pillar_1.2.2      
[13] plyr_1.8.4         stringr_1.3.0      munsell_0.4.3     
[16] gtable_0.2.0       caTools_1.17.1     codetools_0.2-15  
[19] evaluate_0.10.1    knitr_1.20         permute_0.9-4     
[22] flexmix_2.3-14     Rcpp_0.12.16       KernSmooth_2.23-15
[25] scales_0.5.0       backports_1.1.2    gdata_2.18.0      
[28] vegan_2.5-1        maptpx_1.9-3       picante_1.6-2     
[31] gplots_3.0.1       digest_0.6.15      stringi_1.2.2     
[34] grid_3.4.1         rprojroot_1.3-2    cowplot_0.9.2     
[37] tools_3.4.1        bitops_1.0-6       magrittr_1.5      
[40] lazyeval_0.2.1     tibble_1.4.2       cluster_2.0.7-1   
[43] ape_5.1            MASS_7.3-50        SQUAREM_2017.10-1 
[46] rmarkdown_1.9      iterators_1.0.9    boot_1.3-20       
[49] nnet_7.3-12        nlme_3.1-137       git2r_0.21.0      
[52] compiler_3.4.1    </code></pre>
</div>

<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>

<hr>
<p>
    This <a href="http://rmarkdown.rstudio.com">R Markdown</a> site was created with <a href="https://github.com/jdblischak/workflowr">workflowr</a>
</p>
<hr>

<!-- To enable disqus, uncomment the section below and provide your disqus_shortname -->

<!-- disqus
  <div id="disqus_thread"></div>
    <script type="text/javascript">
        /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
        var disqus_shortname = 'rmarkdown'; // required: replace example with your forum shortname

        /* * * DON'T EDIT BELOW THIS LINE * * */
        (function() {
            var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
            dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
            (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
        })();
    </script>
    <noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
    <a href="http://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>
-->


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
