<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="Joseph Paulson" />

<meta name="date" content="2017-12-10" />

<title>Filtering</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-4.5.0/css/font-awesome.min.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}

.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->




<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
  padding-left: 25px;
  text-indent: 0;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">nasalmicrobiome</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jhsiao999/nasalmicrobiome">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Filtering</h1>
<h4 class="author"><em>Joseph Paulson</em></h4>
<h4 class="date"><em>2017-12-10</em></h4>

</div>


<!-- The file analysis/chunks.R contains chunks that define default settings
shared across the workflowr files. -->
<!-- Update knitr chunk options -->
<!-- Insert the date the file was last updated -->
<pre><code>Warning in as.POSIXlt.POSIXct(Sys.time()): unknown timezone &#39;zone/tz/2017c.
1.0/zoneinfo/America/Chicago&#39;</code></pre>
<p><strong>Last updated:</strong> 2017-12-21</p>
<!-- Insert the code version (Git commit SHA1) if Git repository exists and R
 package git2r is installed -->
<p><strong>Code version:</strong> 0c0400c</p>
<hr />
<div id="summary" class="section level2">
<h2>Summary</h2>
<p>We excluded samples and OTUs based on the following criteria:</p>
<p><em>Prior to filtering, there are 42,999 features and 201 samples</em></p>
<ol style="list-style-type: decimal">
<li>Low sequencing depth 1.1. Samples with total count &lt; 1000<br />
1.2. OTUs not present (&gt; 0 read) in any sample<br />
1.3. Satisfies 1.1 AND 1.2</li>
</ol>
<p><em>After Step 1, left with 42,982 features and 197 samples.</em></p>
<ol start="2" style="list-style-type: decimal">
<li>OTUs not present 2.1. Present (&gt; 0 read) in 5 or more samples in 1 2.2. Present and quantified with &gt; 20 read in at least one sample 2.3. Satisfies 2.1 OR 2.2, otherwise excluded.</li>
</ol>
<p><em>After Step 2, left with 5,521 features and 197 samples.</em></p>
<ol start="3" style="list-style-type: decimal">
<li>Excluded OTUs not assigned at the Kindom level</li>
</ol>
<p><em>After Step 3, left with 4,423 features and 197 samples</em></p>
<p>In the end, 4 samples/patients were excluded from the analysis. See the bottom of this page for sample demographics.</p>
<p><span class="math inline">\(~\)</span></p>
<p>Questions:</p>
<ol style="list-style-type: decimal">
<li><p>Maybe included one patient that has ~900 total reads?</p></li>
<li><p>Difference between treatment and control demographics before/after excluded these samples?</p></li>
</ol>
<hr />
</div>
<div id="packages" class="section level2">
<h2>Packages</h2>
<pre class="r"><code>library(metagenomeSeq)</code></pre>
</div>
<div id="loading-data" class="section level2">
<h2>Loading data</h2>
<pre class="r"><code>MRobj = readRDS(&quot;../data/nasal.rds&quot;)
MRobj</code></pre>
<pre><code>MRexperiment (storageMode: environment)
assayData: 42999 features, 201 samples 
  element names: counts 
protocolData: none
phenoData
  sampleNames: EM0042 EM0047 ... E0168 (201 total)
  varLabels: StudyID age ... infnone (94 total)
  varMetadata: labelDescription
featureData
  featureNames: denovo0 denovo1 ... denovo42998 (42999 total)
  fvarLabels: Kingdom Phylum ... Species (7 total)
  fvarMetadata: labelDescription
experimentData: use &#39;experimentData(object)&#39;
Annotation:  </code></pre>
<p><span class="math inline">\(~\)</span></p>
<hr />
</div>
<div id="filter-samples-with-low-depth" class="section level2">
<h2>Filter samples with low depth</h2>
<ul>
<li><p>Filter out samples with total count &lt; 1000.</p></li>
<li><p>Filter out OTUs with total count across samples &lt; 1. In other words, present in at least one sample.</p></li>
</ul>
<pre class="r"><code>hist(log2(colSums(MRobj)),xlab=&quot;Log sequencing depth&quot;,
     main = &quot;Log sequencing depth&quot;)
abline(v=log2(1000))</code></pre>
<p><img src="figure/step1_filtering.Rmd/unnamed-chunk-3-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>MRobj = filterData(MRobj, depth=1000, present=1)
MRobj</code></pre>
<pre><code>MRexperiment (storageMode: environment)
assayData: 42982 features, 197 samples 
  element names: counts 
protocolData: none
phenoData
  sampleNames: EM0042 EM0047 ... E0168 (197 total)
  varLabels: StudyID age ... infnone (94 total)
  varMetadata: labelDescription
featureData
  featureNames: denovo0 denovo1 ... denovo42998 (42982 total)
  fvarLabels: Kingdom Phylum ... Species (7 total)
  fvarMetadata: labelDescription
experimentData: use &#39;experimentData(object)&#39;
Annotation:  </code></pre>
<p><span class="math inline">\(~\)</span></p>
<hr />
</div>
<div id="filter-otus-not-present" class="section level2">
<h2>Filter OTUs not present</h2>
<ul>
<li><p>OTUs present (&gt; 0 read) in 5 or more samples</p></li>
<li><p>OTUs that are quantified with &gt; 20 reads in at least one sample</p></li>
</ul>
<pre class="r"><code>nmat = MRcounts(MRobj)
keep = which( rowSums(nmat&gt;0)&gt;=5 | rowSums(nmat&gt;=20)&gt;0 )
MRobj=MRobj[keep,]</code></pre>
<hr />
</div>
<div id="unassigned" class="section level2">
<h2>Unassigned</h2>
<p>After filtering, we have 5,521 features and 197 samples.</p>
<pre class="r"><code>MRobj</code></pre>
<pre><code>MRexperiment (storageMode: environment)
assayData: 5521 features, 197 samples 
  element names: counts 
protocolData: none
phenoData
  sampleNames: EM0042 EM0047 ... E0168 (197 total)
  varLabels: StudyID age ... infnone (94 total)
  varMetadata: labelDescription
featureData
  featureNames: denovo2 denovo6 ... denovo42992 (5521 total)
  fvarLabels: Kingdom Phylum ... Species (7 total)
  fvarMetadata: labelDescription
experimentData: use &#39;experimentData(object)&#39;
Annotation:  </code></pre>
<p>Number of unique organisms we find within each clade. PS. 2 at the Kindom level due to the unassigned</p>
<pre class="r"><code>sapply(fData(MRobj),function(i)length(unique(i)))</code></pre>
<pre><code>Kingdom  Phylum   Class   Order  Family   Genus Species 
      2      34      79     153     300     544     617 </code></pre>
<p>Number of features unassigned at the Kindom level. (1,098 out of total 5521 OTUs)</p>
<pre class="r"><code>length(which(fData(MRobj)[,1]==&quot;Unassigned&quot;))</code></pre>
<pre><code>[1] 1098</code></pre>
<p>Filtering Unassigned.</p>
<pre class="r"><code>MRobj = MRobj[- which(fData(MRobj)[,1]==&quot;Unassigned&quot;),]
saveRDS(MRobj,file=&quot;../data/nasal_filtered.rds&quot;)</code></pre>
<p>After filtering, we have 4,423 features and 197 samples.</p>
<pre class="r"><code>MRobj</code></pre>
<pre><code>MRexperiment (storageMode: environment)
assayData: 4423 features, 197 samples 
  element names: counts 
protocolData: none
phenoData
  sampleNames: EM0042 EM0047 ... E0168 (197 total)
  varLabels: StudyID age ... infnone (94 total)
  varMetadata: labelDescription
featureData
  featureNames: denovo6 denovo17 ... denovo42992 (4423 total)
  fvarLabels: Kingdom Phylum ... Species (7 total)
  fvarMetadata: labelDescription
experimentData: use &#39;experimentData(object)&#39;
Annotation:  </code></pre>
</div>
<div id="excluded-sample" class="section level2">
<h2>Excluded sample</h2>
<p>4 samples were excluded from the filtered data: EM0062, E0194, EM0088, EM0053.</p>
<pre class="r"><code>pre_filtered &lt;- readRDS(&quot;../data/nasal.rds&quot;)
post_filtered &lt;- readRDS(&quot;../data/nasal_filtered.rds&quot;)

excluded_samples &lt;- setdiff(colnames(MRcounts(pre_filtered)), colnames(MRcounts(post_filtered)))
excluded_samples</code></pre>
<pre><code>[1] &quot;EM0062&quot; &quot;E0194&quot;  &quot;EM0088&quot; &quot;EM0053&quot;</code></pre>
<pre class="r"><code>colSums(MRcounts(pre_filtered))[colnames(pre_filtered) %in% c(&quot;EM0062&quot;, &quot;E0194&quot;, &quot;EM0088&quot;, &quot;EM0053&quot;)]</code></pre>
<pre><code>EM0062  E0194 EM0088 EM0053 
   104    934    271     36 </code></pre>
<p>Phenotypes of the 4 excluded samples.</p>
<p><em>EM0062</em></p>
<ul>
<li><p>Demographics: age 66, male, white, underwent cardiac surgery, studysite = 1, seqrun = 2</p></li>
<li><p>Primary outcome: anyinf6m (infection 6 mo. post surgery) = 1, naswabsa1 (nasal swab culture positive) = 1</p></li>
</ul>
<pre class="r"><code>pData(pre_filtered)[rownames(pData(pre_filtered)) == &quot;EM0062&quot;,]</code></pre>
<pre><code>       StudyID age gender race srgtpop htprop wtprop bmiprop ster3prop
EM0062  EM0062  66      0    0       1  180.3    114    35.1         0
       cadprop chfdgprop htnprop cholprop dbprop smokcuprop cvdprop
EM0062       1         1       1        0      1          0       0
       padprop lungdisprop asthmaprop dialprop cancerprop obesity
EM0062       0           0          0        0          1       1
       creatprop bunprop hgbprop hctprop wbcprop plateprop  wbc  hgb  hct
EM0062      0.91      17    13.7    39.9    6600       237 6400 15.7 46.9
       plate ssi ssidepth SSIorg1 SSIorg2 SSIorg3 infdeath cardsrgsc
EM0062   191   1     deep    none    &lt;NA&gt;    &lt;NA&gt;        0         1
       cransrgsc spinsrgsc vascsrgsc anyinf30 anyinf6m death6m ssideep6m
EM0062         0         0         0        1        1       0         1
       bloodinf6m pneumon6m naswabtp studysite DateExtracted seqrun
EM0062          0         0        3         1    2014-05-12      2
       infabprop miprop smokprprop ulcerprop liverprop readm_30d maristat3
EM0062         0      0          1         0         0         1         1
       naswabsa1 naswabmrsa1 insurstat3 CharlCA Charlcat educat4
EM0062         1           0          2       0        1       1
       nonsysester1prop bronmedprop lungdz medincome hosprop asa CharlF
EM0062                0           0      0     56221       1   4      2
       COPD h2blockprop skinev1prop nasalster antihist allergies outinop
EM0062    0           1           0         0        0         0       1
             dtop ssideepsepi ssideepSA ssideepStaph ssideepstrep
EM0062 09/04/2009           0         0            0            0
       ssideeppseudo ssideepentero ssideepnone infSepi infSA infStaph
EM0062             0             0           1       0     0        0
       infStrep infpseudo infEntero infnone
EM0062        0         0         0       1</code></pre>
<p><span class="math inline">\(~\)</span></p>
<p><em>E0194</em></p>
<ul>
<li><p>Demographics: age 66, male, white, underwent cardiac surgery, study site = 0, seqrun = 2</p></li>
<li><p>Primary outcome: anyinf6m (infection 6 mo. post surgery) = 0, naswabsa1 (nasal swab culture positive) = 0</p></li>
</ul>
<pre class="r"><code>pData(pre_filtered)[rownames(pData(pre_filtered)) == &quot;E0194&quot;,]</code></pre>
<pre><code>      StudyID age gender race srgtpop htprop wtprop bmiprop ster3prop
E0194   E0194  66      0    0       1    165     88    32.3         0
      cadprop chfdgprop htnprop cholprop dbprop smokcuprop cvdprop padprop
E0194       1         0       1        1      0          0       0       0
      lungdisprop asthmaprop dialprop cancerprop obesity creatprop bunprop
E0194           0          0        1          0       1       4.9      28
      hgbprop hctprop wbcprop plateprop  wbc hgb  hct plate ssi ssidepth
E0194     8.8    28.4    3170       207 2600 8.1 25.6   201   0     &lt;NA&gt;
      SSIorg1 SSIorg2 SSIorg3 infdeath cardsrgsc cransrgsc spinsrgsc
E0194    &lt;NA&gt;    &lt;NA&gt;    &lt;NA&gt;        1         1         0         0
      vascsrgsc anyinf30 anyinf6m death6m ssideep6m bloodinf6m pneumon6m
E0194         0        0        0       1         0          0         0
      naswabtp studysite DateExtracted seqrun infabprop miprop smokprprop
E0194        2         0    2013-08-21      2         1      1          1
      ulcerprop liverprop readm_30d maristat3 naswabsa1 naswabmrsa1
E0194         0         0         0         1         0           0
      insurstat3 CharlCA Charlcat educat4 nonsysester1prop bronmedprop
E0194          1       0        3       0                0           0
      lungdz medincome hosprop asa CharlF COPD h2blockprop skinev1prop
E0194      0     41580       1   4      5    0           0           1
      nasalster antihist allergies outinop       dtop ssideepsepi
E0194         0        0         0       2 02/20/2008           0
      ssideepSA ssideepStaph ssideepstrep ssideeppseudo ssideepentero
E0194         0            0            0             0             0
      ssideepnone infSepi infSA infStaph infStrep infpseudo infEntero
E0194           0       0     0        0        0         0         0
      infnone
E0194       0</code></pre>
<p><span class="math inline">\(~\)</span></p>
<p><em>EM0088</em></p>
<ul>
<li><p>Demographics: age 74, female, white, underwent cardiac surgery, studysite = 1, seqrun = 2</p></li>
<li><p>Primary outcome: anyinf6m (infection 6 mo. post surgery) = 1, naswabsa1 (nasal swab culture positive) = 1</p></li>
</ul>
<pre class="r"><code>pData(pre_filtered)[rownames(pData(pre_filtered)) == &quot;EM0088&quot;,]</code></pre>
<pre><code>       StudyID age gender race srgtpop htprop wtprop bmiprop ster3prop
EM0088  EM0088  74      1    0       1    160   65.4    25.5         0
       cadprop chfdgprop htnprop cholprop dbprop smokcuprop cvdprop
EM0088       0         0       1        0      0          1       0
       padprop lungdisprop asthmaprop dialprop cancerprop obesity
EM0088       0           0          0        0          0       0
       creatprop bunprop hgbprop hctprop wbcprop plateprop  wbc  hgb  hct
EM0088      1.73      34    12.5    36.7    6300       156 6000 12.6 36.5
       plate ssi ssidepth      SSIorg1     SSIorg2     SSIorg3 infdeath
EM0088   156   1     deep enterococcus bacteroides pseudomonas        0
       cardsrgsc cransrgsc spinsrgsc vascsrgsc anyinf30 anyinf6m death6m
EM0088         1         0         0         0        1        1       1
       ssideep6m bloodinf6m pneumon6m naswabtp studysite DateExtracted
EM0088         1          1         1        3         1    2014-05-12
       seqrun infabprop miprop smokprprop ulcerprop liverprop readm_30d
EM0088      2         1      0          1         0         0         0
       maristat3 naswabsa1 naswabmrsa1 insurstat3 CharlCA Charlcat educat4
EM0088         1         1           1          1       0        3       0
       nonsysester1prop bronmedprop lungdz medincome hosprop asa CharlF
EM0088                0           1      1     95094       1   4      5
       COPD h2blockprop skinev1prop nasalster antihist allergies outinop
EM0088    1           1           0         0        0         0       1
             dtop ssideepsepi ssideepSA ssideepStaph ssideepstrep
EM0088 07/13/2011           0         0            0            1
       ssideeppseudo ssideepentero ssideepnone infSepi infSA infStaph
EM0088             1             0           0       0     0        0
       infStrep infpseudo infEntero infnone
EM0088        1         1         0       0</code></pre>
<p><span class="math inline">\(~\)</span></p>
<p><em>EM0053</em></p>
<ul>
<li><p>Demographics: age 64, male, white, underwent cardiac surgery, studysite = 1, seqrun = 2</p></li>
<li><p>Primary outcome: anyinf6m (infection 6 mo. post surgery) = 1, naswabsa1 (nasal swab culture positive) = 0</p></li>
</ul>
<pre class="r"><code>pData(pre_filtered)[rownames(pData(pre_filtered)) == &quot;EM0053&quot;,]</code></pre>
<pre><code>       StudyID age gender race srgtpop htprop wtprop bmiprop ster3prop
EM0053  EM0053  64      0    0       1  177.8     94    29.7         0
       cadprop chfdgprop htnprop cholprop dbprop smokcuprop cvdprop
EM0053       1         0       1        0      0          0       0
       padprop lungdisprop asthmaprop dialprop cancerprop obesity
EM0053       0           0          0        0          0       0
       creatprop bunprop hgbprop hctprop wbcprop plateprop  wbc  hgb hct
EM0053      1.17       8    15.5    42.4    5900       202 4200 13.7  38
       plate ssi ssidepth    SSIorg1 SSIorg2 SSIorg3 infdeath cardsrgsc
EM0053     5   1     deep klebsiella proteus    &lt;NA&gt;        0         1
       cransrgsc spinsrgsc vascsrgsc anyinf30 anyinf6m death6m ssideep6m
EM0053         0         0         0        1        1       0         1
       bloodinf6m pneumon6m naswabtp studysite DateExtracted seqrun
EM0053          1         1        3         1    2014-05-12      2
       infabprop miprop smokprprop ulcerprop liverprop readm_30d maristat3
EM0053         1      1          0         0         0         0         0
       naswabsa1 naswabmrsa1 insurstat3 CharlCA Charlcat educat4
EM0053         0           0          2       0        1       1
       nonsysester1prop bronmedprop lungdz medincome hosprop asa CharlF
EM0053                0           0      0     92509       0   4      1
       COPD h2blockprop skinev1prop nasalster antihist allergies outinop
EM0053    0           1           1         0        0         0       1
             dtop ssideepsepi ssideepSA ssideepStaph ssideepstrep
EM0053 07/14/2009           0         0            0            0
       ssideeppseudo ssideepentero ssideepnone infSepi infSA infStaph
EM0053             0             1           0       0     0        0
       infStrep infpseudo infEntero infnone
EM0053        0         0         1       0</code></pre>
<hr />
</div>
<div id="session-information" class="section level2">
<h2>Session Information</h2>
<pre><code>R version 3.4.1 (2017-06-30)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS High Sierra 10.13

Matrix products: default
BLAS: /Library/Frameworks/R.framework/Versions/3.4/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.4/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] parallel  stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
[1] metagenomeSeq_1.20.1 RColorBrewer_1.1-2   glmnet_2.0-13       
[4] foreach_1.4.4        Matrix_1.2-12        limma_3.34.4        
[7] Biobase_2.38.0       BiocGenerics_0.24.0 

loaded via a namespace (and not attached):
 [1] Rcpp_0.12.14       knitr_1.17         magrittr_1.5      
 [4] lattice_0.20-35    stringr_1.2.0      caTools_1.17.1    
 [7] tools_3.4.1        grid_3.4.1         KernSmooth_2.23-15
[10] git2r_0.20.0       gtools_3.5.0       htmltools_0.3.6   
[13] iterators_1.0.9    matrixStats_0.52.2 yaml_2.1.16       
[16] rprojroot_1.2      digest_0.6.13      bitops_1.0-6      
[19] codetools_0.2-15   evaluate_0.10.1    rmarkdown_1.8     
[22] gdata_2.18.0       stringi_1.1.6      compiler_3.4.1    
[25] gplots_3.0.1       backports_1.1.2   </code></pre>
</div>

<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>

<hr>
<p>
    This <a href="http://rmarkdown.rstudio.com">R Markdown</a> site was created with <a href="https://github.com/jdblischak/workflowr">workflowr</a>
</p>
<hr>

<!-- To enable disqus, uncomment the section below and provide your disqus_shortname -->

<!-- disqus
  <div id="disqus_thread"></div>
    <script type="text/javascript">
        /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
        var disqus_shortname = 'rmarkdown'; // required: replace example with your forum shortname

        /* * * DON'T EDIT BELOW THIS LINE * * */
        (function() {
            var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
            dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
            (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
        })();
    </script>
    <noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
    <a href="http://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>
-->


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
