<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="JNP" />


<title>QIIME to MRexperiment object</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-4.5.0/css/font-awesome.min.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}

.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->




<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
  padding-left: 25px;
  text-indent: 0;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">nasalmicrobiome</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jhsiao999/nasalmicrobiome">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">QIIME to MRexperiment object</h1>
<h4 class="author"><em>JNP</em></h4>

</div>


<!-- The file analysis/chunks.R contains chunks that define default settings
shared across the workflowr files. -->
<!-- Update knitr chunk options -->
<!-- Insert the date the file was last updated -->
<pre><code>Warning in as.POSIXlt.POSIXct(Sys.time()): unknown timezone &#39;zone/tz/2017c.
1.0/zoneinfo/America/Chicago&#39;</code></pre>
<p><strong>Last updated:</strong> 2017-12-21</p>
<!-- Insert the code version (Git commit SHA1) if Git repository exists and R
 package git2r is installed -->
<p><strong>Code version:</strong> d436804</p>
<hr />
<div id="introductionsummary" class="section level2">
<h2>Introduction/Summary</h2>
<p>Here we combine QIIME output, OTU taxonomy labels, and sample phenotype labels into a <code>metagenomeSeq</code> <code>MRexperiment</code> object <code>nasal.rds</code>. The sources of data are</p>
<ul>
<li><p>QIIME output: <code>otu_table.biom</code></p></li>
<li><p>Taxonomy labels: <code>GSCID_Seqs_4Nauder_rep_set_tax_assignments.txt</code></p></li>
<li><p>Sample phenotype labels: <code>metadata.tsv</code></p></li>
</ul>
<p>Raw data consists of 42,999 features (OTUs) from 201 samples. The OTUs are annotated at each taxonomy level: Kindom, Phylum, Class, Order, Family, Genus, Species, and labeled as unassigned when annotations are not available. There are 94 sample phenotypes, including primary and secondary outcomes.</p>
<p><span class="math inline">\(~\)</span></p>
<hr />
</div>
<div id="setting-up" class="section level2">
<h2>Setting up</h2>
<pre class="r"><code>library(biomformat)
library(metagenomeSeq)
library(dplyr)</code></pre>
<hr />
</div>
<div id="reading-in-the-counts" class="section level2">
<h2>Reading in the counts</h2>
<p>Import QIIME output using <code>biomformat</code>, and convert to an <code>MRexperiment</code> object using <code>metagenomeSeq</code>.</p>
<pre class="r"><code>counts &lt;- read_biom(&quot;../data/otu_table.biom&quot;)
counts = biom2MRexperiment(counts)
cnames = sapply(strsplit(colnames(counts),&quot;\\.&quot;),function(i)ifelse(length(i)&gt;1,i[2],i[1]))
colnames(counts) = cnames</code></pre>
<p>Raw data consists of 42,999 features and 201 samples.</p>
<pre class="r"><code>dim(counts)</code></pre>
<pre><code>Features  Samples 
   42999      201 </code></pre>
<hr />
</div>
<div id="taxonomy" class="section level2">
<h2>Taxonomy</h2>
<p>Read in the QIIME taxonomy and attach to the <code>MRexperiment</code> object. It requires a bit of cleanup due to the file structure.</p>
<pre class="r"><code>fd = read.csv(&quot;../data/GSCID_Seqs_4Nauder_rep_set_tax_assignments.txt&quot;,sep=&quot;;&quot;,
              stringsAsFactors=FALSE,header=FALSE)
species = sapply(strsplit(fd[,7],&quot;\t&quot;),function(i)i[1]) %&gt;% ifelse(is.na(.),yes=&quot; s__&quot;,no=.)
fd[,7] = species
rnames = sapply(strsplit(fd[,1],&quot;\t&quot;),function(i)i[1])
rownames(fd) = rnames
fd[,1] = sapply(strsplit(fd[,1],&quot;\t&quot;),function(i)i[2]) %&gt;% ifelse(is.na(.),yes=&quot; Unassigned&quot;,no=.)
colnames(fd) = c(&quot;Kingdom&quot;,&quot;Phylum&quot;,&quot;Class&quot;,&quot;Order&quot;,&quot;Family&quot;,&quot;Genus&quot;,&quot;Species&quot;)

for(i in seq(fd)){
  fd[,i] = gsub(&quot; &quot;,&quot;&quot;,fd[,i])
}
for(i in seq(fd)){
  lvl = tolower(strsplit(colnames(fd)[i],&quot;&quot;)[[1]][1])
  fd[which(is.na(fd[,i]) | fd[,i]==&quot;&quot;),i] = paste(lvl,&quot;__&quot;,sep=&quot;&quot;)
}

for(i in rev(seq(fd)[-1])){
  lvl = paste(tolower(strsplit(colnames(fd)[i],&quot;&quot;)[[1]][1]),&quot;__&quot;,sep=&quot;&quot;)
  if(any(fd[,i] == lvl)){
    nrows = which(fd[,i] == lvl)
    feel  = fd[nrows,seq(i)]
    feel  = sapply(seq(nrow(feel)),function(j)paste(feel[j,seq(i)],collapse=&quot;;&quot;))
    fd[nrows,i] = feel
  }
}
for(i in seq(fd)) fd[grep(&quot;Unassigned&quot;,fd[,i]),i] = &quot;Unassigned&quot;
fd = fd[match(rownames(counts),rownames(fd)),]
fData(counts) = fd</code></pre>
<p><span class="math inline">\(~\)</span></p>
<hr />
</div>
<div id="metadata" class="section level2">
<h2>Metadata</h2>
<p>Next we read in the metadata (sample phenotype labels) and attach to the <code>MRexperiment</code> object</p>
<pre class="r"><code>pd = read.csv(&quot;../data/metadata.tsv&quot;,sep = &quot;\t&quot;,stringsAsFactors=FALSE)
rownames(pd) = pd[,1]
pd = pd[match(colnames(counts),rownames(pd)),]
pData(counts) = pd</code></pre>
<p>There are 94 phenotype labels for the 201 samples.</p>
<pre class="r"><code>dim(pd)</code></pre>
<pre><code>[1] 201  94</code></pre>
<hr />
</div>
<div id="saving-the-file" class="section level2">
<h2>Saving the file</h2>
<p>Lastly, we save the object as an RDS file.</p>
<pre class="r"><code>saveRDS(counts, file=&quot;../data/nasal.rds&quot;)</code></pre>
<hr />
</div>
<div id="sessioninfo" class="section level2">
<h2>SessionInfo</h2>
<pre><code>R version 3.4.1 (2017-06-30)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS High Sierra 10.13

Matrix products: default
BLAS: /Library/Frameworks/R.framework/Versions/3.4/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.4/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] parallel  stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] dplyr_0.7.4          metagenomeSeq_1.20.1 RColorBrewer_1.1-2  
 [4] glmnet_2.0-13        foreach_1.4.4        Matrix_1.2-12       
 [7] limma_3.34.4         Biobase_2.38.0       BiocGenerics_0.24.0 
[10] biomformat_1.6.0    

loaded via a namespace (and not attached):
 [1] Rcpp_0.12.14       bindr_0.1          compiler_3.4.1    
 [4] git2r_0.20.0       plyr_1.8.4         bitops_1.0-6      
 [7] iterators_1.0.9    tools_3.4.1        zlibbioc_1.24.0   
[10] digest_0.6.13      tibble_1.3.4       jsonlite_1.5      
[13] evaluate_0.10.1    rhdf5_2.22.0       lattice_0.20-35   
[16] pkgconfig_2.0.1    rlang_0.1.4        yaml_2.1.16       
[19] bindrcpp_0.2       stringr_1.2.0      knitr_1.17        
[22] gtools_3.5.0       caTools_1.17.1     rprojroot_1.2     
[25] grid_3.4.1         glue_1.2.0         R6_2.2.2          
[28] rmarkdown_1.8      gdata_2.18.0       magrittr_1.5      
[31] backports_1.1.2    gplots_3.0.1       codetools_0.2-15  
[34] htmltools_0.3.6    matrixStats_0.52.2 assertthat_0.2.0  
[37] KernSmooth_2.23-15 stringi_1.1.6     </code></pre>
</div>

<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>

<hr>
<p>
    This <a href="http://rmarkdown.rstudio.com">R Markdown</a> site was created with <a href="https://github.com/jdblischak/workflowr">workflowr</a>
</p>
<hr>

<!-- To enable disqus, uncomment the section below and provide your disqus_shortname -->

<!-- disqus
  <div id="disqus_thread"></div>
    <script type="text/javascript">
        /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
        var disqus_shortname = 'rmarkdown'; // required: replace example with your forum shortname

        /* * * DON'T EDIT BELOW THIS LINE * * */
        (function() {
            var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
            dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
            (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
        })();
    </script>
    <noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
    <a href="http://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>
-->


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
