<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="Joseph Paulson" />

<meta name="date" content="2017-12-10" />

<title>Assigning Taxonomy and Filtering</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-4.5.0/css/font-awesome.min.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}

.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->




<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
  padding-left: 25px;
  text-indent: 0;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">nasalmicrobiome</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jhsiao999/nasalmicrobiome">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Assigning Taxonomy and Filtering</h1>
<h4 class="author"><em>Joseph Paulson</em></h4>
<h4 class="date"><em>2017-12-10</em></h4>

</div>


<!-- The file analysis/chunks.R contains chunks that define default settings
shared across the workflowr files. -->
<!-- Update knitr chunk options -->
<!-- Insert the date the file was last updated -->
<pre><code>Warning in as.POSIXlt.POSIXct(Sys.time()): unknown timezone &#39;zone/tz/2017c.
1.0/zoneinfo/America/Chicago&#39;</code></pre>
<p><strong>Last updated:</strong> 2017-12-21</p>
<!-- Insert the code version (Git commit SHA1) if Git repository exists and R
 package git2r is installed -->
<p><strong>Code version:</strong> 0cb4508</p>
<hr />
<div id="loading-packages" class="section level2">
<h2>Loading packages</h2>
<pre class="r"><code>library(metagenomeSeq)</code></pre>
<hr />
</div>
<div id="data" class="section level2">
<h2>Data</h2>
<p>Take output from QIIME</p>
<hr />
</div>
<div id="assign-taxonomy" class="section level2">
<h2>Assign taxonomy</h2>
<p>Load data (output from <code>QIMME</code>)</p>
<pre class="r"><code>set.seed(1)
load(&quot;../data/rdata/nasal_2_23_17.rdata&quot;)</code></pre>
<p>Cleaning up QIIME output.</p>
<ul>
<li><p>Convert feature names to chararacter.</p></li>
<li><p>For each OTU, convert <code>NA</code> to <code>Unassigned</code></p></li>
</ul>
<pre class="r"><code>colnames(fData(MRobj)) &lt;- c(&quot;Kingdom&quot;,&quot;Phylum&quot;,&quot;Class&quot;,&quot;Order&quot;,&quot;Family&quot;,&quot;Genus&quot;,&quot;Species&quot;)
for(i in seq(fData(MRobj))) fData(MRobj)[,i] = as.character(fData(MRobj)[,i])

fd &lt;- fData(MRobj)
for(i in seq(fd)){
  lvl = tolower(strsplit(colnames(fd)[i],&quot;&quot;)[[1]][1])
  fd[which(is.na(fd[,i]) | fd[,i]==&quot;&quot;),i] = paste(lvl,&quot;__&quot;,sep=&quot;&quot;)
}
for(i in rev(seq(fd))){
  lvl = paste(tolower(strsplit(colnames(fd)[i],&quot;&quot;)[[1]][1]),&quot;__&quot;,sep=&quot;&quot;)
  if(any(fd[,i]==&quot;&quot; | fd[,i] == lvl)){
    nrows = which(fd[,i]==&quot;&quot; | fd[,i] == lvl)
    feel  = fd[nrows,seq(i)]
    feel  = sapply(seq(nrow(feel)),function(j)paste(feel[j,seq(i)],collapse=&quot;;&quot;))
    fData(MRobj)[nrows,i] = feel
  }
}

for(i in seq(fData(MRobj))) fData(MRobj)[,i] = as.character(fData(MRobj)[,i])
for(i in seq(fData(MRobj))) fData(MRobj)[grep(&quot;Unassigned&quot;,fData(MRobj)[,i]),i] = &quot;Unassigned&quot;</code></pre>
<p><span class="math inline">\(~\)</span></p>
<hr />
</div>
<div id="filter-samples-with-low-depth" class="section level2">
<h2>Filter samples with low depth</h2>
<p><img src="figure/taxonomy-n-filtering.Rmd/unnamed-chunk-4-1.png" width="672" style="display: block; margin: auto;" /></p>
<p><span class="math inline">\(~\)</span></p>
<ul>
<li><p>Filter out samples with total count &lt; 1000.</p></li>
<li><p>Filter out OTUs with total count across samples &lt; 1. In other words, present in at least one sample.</p></li>
</ul>
<pre class="r"><code>MRobj = filterData(MRobj,depth=1000,present=1)</code></pre>
<p><span class="math inline">\(~\)</span></p>
<hr />
</div>
<div id="filter-otus-not-present" class="section level2">
<h2>Filter OTUs not present</h2>
<p><span class="math inline">\(~\)</span></p>
<ul>
<li><p>OTUs present (&gt; 0 read) in 5 or more samples</p></li>
<li><p>OTUs that are quantified with &gt; 20 reads in at least one sample</p></li>
</ul>
<pre class="r"><code>nmat &lt;- MRcounts(MRobj)
keep &lt;- which( rowSums(nmat&gt;0)&gt;=5 | rowSums(nmat&gt;=20)&gt;0 )
MRobj &lt;- MRobj[keep,]</code></pre>
<p><span class="math inline">\(~\)</span></p>
<p>After filtering, we have 5,521 features and 197 samples.</p>
<pre class="r"><code>MRobj</code></pre>
<pre><code>MRexperiment (storageMode: environment)
assayData: 5521 features, 197 samples 
  element names: counts 
protocolData: none
phenoData
  rowNames: 1 2 ... 201 (197 total)
  varLabels: StudyID age ... infnone (94 total)
  varMetadata: labelDescription
featureData
  featureNames: denovo2 denovo6 ... denovo42992 (5521 total)
  fvarLabels: Kingdom Phylum ... Species (7 total)
  fvarMetadata: labelDescription
experimentData: use &#39;experimentData(object)&#39;
Annotation:  </code></pre>
<p><span class="math inline">\(~\)</span></p>
<p>Display number of unique organisms we find within each clade. ps. 2 kingdom is due to the unassigned</p>
<pre class="r"><code>sapply(fData(MRobj),function(i)length(unique(i)))</code></pre>
<pre><code>Kingdom  Phylum   Class   Order  Family   Genus Species 
      2      31      76     145     264     505     605 </code></pre>
<p><span class="math inline">\(~\)</span></p>
<p>Number of features unassigned at the Kindom level. (1098 out of total 5521 OTUs)</p>
<pre class="r"><code>length(which(fData(MRobj)[,1]==&quot;Unassigned&quot;))</code></pre>
<pre><code>[1] 1098</code></pre>
<p><span class="math inline">\(~\)</span></p>
<p>save data to a rdata file.</p>
<pre class="r"><code>save(MRobj, file = &quot;../data/data-filtered.rdata&quot;)</code></pre>
<hr />
</div>
<div id="session-information" class="section level2">
<h2>Session information</h2>
<pre><code>R version 3.4.1 (2017-06-30)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS High Sierra 10.13

Matrix products: default
BLAS: /Library/Frameworks/R.framework/Versions/3.4/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.4/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] parallel  stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
[1] metagenomeSeq_1.20.0 RColorBrewer_1.1-2   glmnet_2.0-13       
[4] foreach_1.4.3        Matrix_1.2-12        limma_3.34.2        
[7] Biobase_2.38.0       BiocGenerics_0.24.0 

loaded via a namespace (and not attached):
 [1] Rcpp_0.12.14       knitr_1.17         magrittr_1.5      
 [4] lattice_0.20-35    stringr_1.2.0      caTools_1.17.1    
 [7] tools_3.4.1        grid_3.4.1         KernSmooth_2.23-15
[10] git2r_0.19.0       gtools_3.5.0       htmltools_0.3.6   
[13] iterators_1.0.8    matrixStats_0.52.2 yaml_2.1.15       
[16] rprojroot_1.2      digest_0.6.12      bitops_1.0-6      
[19] codetools_0.2-15   evaluate_0.10.1    rmarkdown_1.8     
[22] gdata_2.18.0       stringi_1.1.6      compiler_3.4.1    
[25] gplots_3.0.1       backports_1.1.1   </code></pre>
</div>

<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>

<hr>
<p>
    This <a href="http://rmarkdown.rstudio.com">R Markdown</a> site was created with <a href="https://github.com/jdblischak/workflowr">workflowr</a>
</p>
<hr>

<!-- To enable disqus, uncomment the section below and provide your disqus_shortname -->

<!-- disqus
  <div id="disqus_thread"></div>
    <script type="text/javascript">
        /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
        var disqus_shortname = 'rmarkdown'; // required: replace example with your forum shortname

        /* * * DON'T EDIT BELOW THIS LINE * * */
        (function() {
            var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
            dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
            (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
        })();
    </script>
    <noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
    <a href="http://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>
-->


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
