---
title: "Compute variance explained under different K"
author: "Joyce Hsiao"
output: html_document
---

<!-- The file analysis/chunks.R contains chunks that define default settings
shared across the workflowr files. -->
```{r read-chunk, include=FALSE, cache=FALSE}
knitr::read_chunk("chunks.R")
```

<!-- Update knitr chunk options -->
```{r knitr-opts-chunk, include=FALSE}
```

<!-- Insert the date the file was last updated -->
```{r last-updated, echo=FALSE, results='asis'}
```

<!-- Insert the code version (Git commit SHA1) if Git repository exists and R
 package git2r is installed -->
```{r code-version, echo=FALSE, results='asis'}
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning=F, message=F)
```

## Background and summary

We compare fitting under differet k and computed for every feature, sample variance in abundance that is explained by sample cluster membership. 

Results showed that k=2 gives the largest variance explained (median of sample variances is higher than the other k).

---

## Analysis

```{r}
library(ggplot2)
library(metagenomeSeq)
```

Load data

```{r}
MRobj <- readRDS("../data/nasal_filtered_normed_batchcorrected.rds")
counts <- MRcounts(MRobj,norm=FALSE,log=FALSE)
```

Load previously computed results.

```{r}
clust_2 <- readRDS("../output/gom-k2-best-seed-67.rds")
clust_3 <- readRDS("../output/gom-k3-best-seed-67.rds")
clust_4 <- readRDS("../output/gom-k4-best-seed-59.rds")
clust_5 <- readRDS("../output/gom-k5-best-seed-307.rds")
fits <- c(clust_2, clust_3, clust_4, clust_5)
names(fits) <- c("clust_2", "clust_3", "clust_4", "clust_5")
```

Extract sampel cluster membership.

```{r}
membership <- lapply(fits, function(xx) {
     apply(xx$omega, 1, which.max)
  })
names(membership) <- c("clust_2", "clust_3", "clust_4", "clust_5")
```

Compute variance explained.

```{r, eval = FALSE}
varprop_species <- lapply(1:length(membership), function(index) {
  which_sample <- which(colnames(counts) %in% names(membership[[index]]))
  counts_which <- counts[, which_sample]
  var <- sapply(1:nrow(counts_which), function(index2) {
            res <- lm(log2(counts_which[index2,]+2) ~ membership[[index]])
            sumOfsquares <- anova(res)[[2]]
            sumOfsquares[1]/sum(sumOfsquares) })
  return(var)
})
names(varprop_species) <- names(membership)
saveRDS(varprop_species, file = "../data/varprop.rds")
```

Plot variance explained

```{r}
varprop <- readRDS("../data/varprop.rds")
varprop_long <- data.frame(
  prop = do.call(c, varprop),
  cluster = factor(rep(2:5, each = nrow(varprop))) )
varprop_long$cluster <- as.numeric(as.character(varprop_long$cluster))
```

K = 2 to 5

```{r}
ggplot(varprop_long, aes(x = factor(cluster), y = prop,
                            col = factor(cluster)), size = .2) +
  geom_violin(trim = FALSE, size=.5) +
  geom_boxplot(width=0.1, size = .6) +
  labs(x = "Number of clusters", y = "Proportion variance explained")
```

---

## Compare variances explained

Sig difference between K=2 vs K=4,5, but no sig. difference between K=2 and K=3.

```{r}
wilcox.test(varprop$clust_2, varprop$clust_3)
wilcox.test(varprop$clust_2, varprop$clust_4)
wilcox.test(varprop$clust_2, varprop$clust_5)
```

Consider within-sample diversity difference for when K=3.

```{r}
library(data.table)
shannon_values <- fread("~/Dropbox/current-projects/nasalMicrobiomeManuscript/draft-results/shannon_values.tsv")

all.equal(shannon_values$V1, rownames(fits$clust_3$omega))

shannon_values$GOM <- sapply(1:nrow(fits$clust_3$omega), function(i) {
  which.max(fits$clust_3$omega[i,]) })

boxplot(shannon_values$V2~ shannon_values$GOM)

levels(shannon_values$GOM) <- c("Group 1", "Group 2", "Group 3")
library(ggplot2)
ggplot(shannon_values, aes(x=GOM, y = V2, fill = factor(GOM))) +
  geom_violin(col = "gray50") +
  scale_fill_manual(values = c(brewer.pal(12,"Set3")[1],
                                 brewer.pal(12,"Set3")[2],
                                 brewer.pal(12,"Set3")[3])) +
  geom_boxplot(width=.25) + ylab("Shannon diversity") +
  scale_x_discrete(name = "Cluster membership",
                   labels = c("Group 1", "Group 2", "Group 3")) +
  guides(fill=guide_legend(title="")) +
  ggtitle("Alpha diversity of K=3")
```


Consider within-sample diversity difference for when K=4.

```{r}
library(data.table)
shannon_values <- fread("~/Dropbox/current-projects/nasalMicrobiomeManuscript/draft-results/shannon_values.tsv")

all.equal(shannon_values$V1, rownames(fits$clust_4$omega))

shannon_values$GOM <- sapply(1:nrow(fits$clust_4$omega), function(i) {
  which.max(fits$clust_4$omega[i,]) })

levels(shannon_values$GOM) <- c("Group 1", "Group 2", "Group 3", "Group 4")
library(ggplot2)
ggplot(shannon_values, aes(x=GOM, y = V2, fill = factor(GOM))) +
  geom_violin(col = "gray50") +
  scale_fill_manual(values = c(brewer.pal(12,"Set3")[1],
                                 brewer.pal(12,"Set3")[2],
                                 brewer.pal(12,"Set3")[3],
                                 brewer.pal(12,"Set3")[4])) +
  geom_boxplot(width=.25) + ylab("Shannon diversity") +
  scale_x_discrete(name = "Cluster membership",
                   labels = c("Group 1", "Group 2", "Group 3", "Group 4")) +
  guides(fill=guide_legend(title="")) +
  ggtitle("Alpha diversity of K=4")
```


Consider within-sample diversity difference for when K=5.

```{r}
library(data.table)
shannon_values <- fread("~/Dropbox/current-projects/nasalMicrobiomeManuscript/draft-results/shannon_values.tsv")

all.equal(shannon_values$V1, rownames(fits$clust_5$omega))

shannon_values$GOM <- sapply(1:nrow(fits$clust_5$omega), function(i) {
  which.max(fits$clust_5$omega[i,]) })

boxplot(shannon_values$V2~ shannon_values$GOM)

levels(shannon_values$GOM) <- c("Group 1", "Group 2", "Group 3", "Group 4", "Group 5")
library(ggplot2)
ggplot(shannon_values, aes(x=GOM, y = V2, fill = factor(GOM))) +
  geom_violin(col = "gray50") +
  scale_fill_manual(values = c(brewer.pal(12,"Set3")[1],
                                 brewer.pal(12,"Set3")[2],
                                 brewer.pal(12,"Set3")[3],
                                 brewer.pal(12,"Set3")[4],
                                 brewer.pal(12,"Set3")[5])) +
  geom_boxplot(width=.25) + ylab("Shannon diversity") +
  scale_x_discrete(name = "Cluster membership",
                   labels = c("Group 1", "Group 2", "Group 3", "Group 4", "Group 5")) +
  guides(fill=guide_legend(title="")) +
  ggtitle("Alpha diversity of K=5")
```

---

## Consider when K=2, the taxa that has high prob of variance explained

```{r}
rownames(counts)[order(varprop$clust_2, decreasing = F)[1:10]]

library(data.table)
de_results <- fread("~/Dropbox/current-projects/nasalMicrobiomeManuscript/draft-results/tbl.tsv")

all.equal(de_results$V1[match(rownames(counts), de_results$V1)], rownames(counts))
de_results <- de_results[match(rownames(counts), de_results$V1),]

plot(varprop$clust_2,-log10(de_results$adjPvalue),
     xlab = "Proportion of variance explained of K=2",
     ylab="-log10(adjusted P-value)",
     pch=16, cex=.7,
     col=(de_results$adjPvalue<.01)+1)
abline(h=2,lty=2,col="black")

de_results$varprop_clust2 <- varprop$clust_2
```

Append the variance proportion explained to the FDR results.

```{r, eval=F}
de_results <- de_results[order(de_results$adjPvalues),]
write.table(de_results,
            "~/Dropbox/current-projects/nasalMicrobiomeManuscript/draft-results/tbl_updated.tsv",
            quote=F, col.names = NA, sep="\t")

#tmp <- fread("~/Dropbox/current-projects/nasalMicrobiomeManuscript/draft-results/tbl_updated.tsv")
```


---

## Session information

```{r, echo = F}
sessionInfo()
```

