---
title: "Applying GoM to HMP data"
author: "Joyce Hsiao"
output: html_document
---

<!-- The file analysis/chunks.R contains chunks that define default settings
shared across the workflowr files. -->
```{r read-chunk, include=FALSE, cache=FALSE}
#knitr::read_chunk("chunks.R")
```

<!-- Update knitr chunk options -->
```{r knitr-opts-chunk, include=FALSE}
```

<!-- Insert the date the file was last updated -->
```{r last-updated, echo=FALSE, results='asis'}
```

<!-- Insert the code version (Git commit SHA1) if Git repository exists and R
 package git2r is installed -->
```{r code-version, echo=FALSE, results='asis'}
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning=F, message=F)
```

---

## Objectives

* Applying GoM to HMP data. 

* Compare K=2-5 GoM profile and Taxa profile with that in our data.


---

## Results

* Run GoM on HMP anterio nares data 389 samples and 14,785 OTUs. After filtering (for features and samples), we have 1,506 OTUs OTUs and 370 samples. 

* Read count corrected for sequencing run center at the Species level.

* Run K=2-5.


---

## HMP data analysis

* Data curated by JNP.

* Analyzed 389 samples collecetd from "anterior nares". 

* Run center is known to be associated with batch effects.



### Filtering

```{r}
library(metagenomeSeq)
library(CountClust)
```

```{r, eval=F}
library(metagenomeSeq)
load("../data/hmp.rdata")

MRobj <- hmp[,pData(hmp)$HMPbodysupersite == "Airways"]

# filtering
MRobj_count <- MRcounts(MRobj, norm=F, log=F)

# keep features present (>0 read) in at least 1% of the samples
keep_rowmeans <- rowSums(MRobj_count > 0) >= ncol(MRobj_count)*.01

# keep features > 5 read in at least one sample
summary(c(MRobj_count))
keep_reads <- rowSums(MRobj_count > 5) > 0

keep_features <- keep_rowmeans | keep_reads

# filtering data
MRobj_filtered <- MRobj[keep_features,]

# filter sample
keep_sample <- colMeans(MRcounts(MRobj_filtered)>0) > .01
MRobj_filtered <- MRobj_filtered[,keep_sample]

dim(MRobj_filtered)
```

### Batch correction

```{r, eval=F}
obj <- aggTax(MRobj_filtered,lvl="species")
obj@expSummary$expSummary$normFactors <- normFactors(MRobj_filtered)

counts <- t(MRcounts(obj,norm=FALSE,log=FALSE))
batchcorrect_counts <- BatchCorrectedCounts(counts,
                                         batch_lab = factor(pData(MRobj_filtered)$RUNCENTER),
                                         use_parallel=FALSE)
counts = t(batchcorrect_counts)

assayData(obj)[["counts"]] = counts
#obj
```

Output filtered, normalized, batch-corrected counts.

```{r, eval=F}
saveRDS(obj, file="../data/hmp_processed.rds")
```

---

### GoM clustering

Try 100 random seeds. 

```{r, eval=F}
obj <- readRDS(file="../data/hmp_processed.rds")

library(CountClust)


# fit GoM
MRobj_species <- aggTax(obj,lvl="species")
counts <- MRcounts(MRobj_species,norm=FALSE,log=FALSE)
clust_2 <- FitGoM(t(counts), K=c(2), tol=0.00001, num_trials = 100)
clust_3 <- FitGoM(t(counts), K=c(3), tol=0.00001, num_trials = 100)
clust_4 <- FitGoM(t(counts), K=c(4), tol=0.00001, num_trials = 100)
clust_5 <- FitGoM(t(counts), K=c(5), tol=0.00001, num_trials = 100)
# clust_6 <- FitGoM(t(counts), K=c(6), tol=0.00001, num_trials = 100)
# clust_7 <- FitGoM(t(counts), K=c(7), tol=0.00001, num_trials = 100)
# clust_8 <- FitGoM(t(counts), K=c(8), tol=0.00001, num_trials = 100)
# clust_9 <- FitGoM(t(counts), K=c(9), tol=0.00001, num_trials = 100)
# clust_10 <- FitGoM(t(counts), K=c(10), tol=0.00001, num_trials = 100)

fits <- list(clust_2$fit, clust_3$fit, clust_4$fit, clust_5$fit)
membership <- lapply(fits, function(xx) {
     apply(xx$omega, 1, which.max)
  })
names(membership) <- paste0("clust_", c(2:5))

saveRDS(fits, file = "../output/hmp.Rmd/hmp_clusters.rds")
saveRDS(membership, file = "../output/hmp.Rmd/hmp_clusters_membership.rds")
```

GoM results

```{r}
fits <- readRDS("../output/hmp.Rmd/hmp_clusters.rds")
membership <- readRDS("../output/hmp.Rmd/hmp_clusters_membership.rds")
# log bayes keeps increasing
plot(y=sapply(fits, function(x) x$BF),
     x=as.integer(c(2:5), pch=16),
     ylab="log Bayes Factor", axes = F, xlab="Clusters")
axis(1, at=c(1:6))
axis(2)


table(membership$clust_2)
table(membership$clust_3)
table(membership$clust_4)
table(membership$clust_5)


fits <- readRDS("../output/hmp.Rmd/hmp_clusters.rds")
membership <- readRDS("../output/hmp.Rmd/hmp_clusters_membership.rds")

pdf("../output/hmp.Rmd/hmp_clusts_structure.pdf")
for (i in 1:length(fits)) {
  annotation <- data.frame(
  sample_id = paste0("X", c(1:NROW(fits[[i]]$omega))),
  tissue_label = factor(membership[[i]],
                        levels = as.character(1:(i+1)) ) ) 
  rownames(fits[[i]]$omega) <- annotation$sample_id
  print(StructureGGplot(omega = fits[[i]]$omega,
                  annotation = annotation,
                  palette = RColorBrewer::brewer.pal(i+1, "Accent"),
                  yaxis_label = "Group",
                  order_sample = TRUE,
                  axis_tick = list(axis_ticks_length = .1,
                                   axis_ticks_lwd_y = .1,
                                   axis_ticks_lwd_x = .1,
                                   axis_label_size = 7,
                                   axis_label_face = "bold")) )
}
dev.off()
```


Taxonomical profile

```{r, eval=F}
library(reshape2)

obj <- readRDS(file="../data/hmp_processed.rds")

# code from code/figs-paulson-04212018.R
pdf("../output/hmp.Rmd/hmp_clusts_taxa.pdf")

pp <- plotBar(obj,lvl='genus',cl=membership$clust_2,n=20)

getPalette = colorRampPalette(brewer.pal(9, "Set1"))
cols <- getPalette(21)
plotBar.aggregate(obj[,order(colnames(obj))],lvl='genus',n=20,
                  plot.factor_levels = levels(pp$dd$variable))$p +
  scale_fill_manual(values = cols) + xlab('Samples') +
  guides(fill=guide_legend(ncol=1))


getPalette = colorRampPalette(brewer.pal(9, "Set1"))
plotBar(obj,lvl='genus',cl=membership$clust_2,n=20)$p +
  scale_fill_manual(values = getPalette(21)) +
  guides(fill=guide_legend(ncol=1))
  

getPalette = colorRampPalette(brewer.pal(9, "Set1"))
plotBar(obj,lvl='genus',cl=membership$clust_3,n=20)$p +
  scale_fill_manual(values = getPalette(21)) +
  guides(fill=guide_legend(ncol=1))

getPalette = colorRampPalette(brewer.pal(9, "Set1"))
plotBar(obj,lvl='genus',cl=membership$clust_4,n=20)$p +
  scale_fill_manual(values = getPalette(21)) +
  guides(fill=guide_legend(ncol=1))

getPalette = colorRampPalette(brewer.pal(9, "Set1"))
plotBar(obj,lvl='genus',cl=membership$clust_5,n=20)$p +
  scale_fill_manual(values = getPalette(21)) +
  guides(fill=guide_legend(ncol=1))
  
dev.off()
```


---


## Current data

See [Grades of membership model finalized](step3_gradesOfMembership.html) for the results.


```{r, eval=F}
counts_clust <- readRDS(file = "../data/count-clust.rds")


MRobj = readRDS("../data/nasal_filtered_normed_batchcorrected.rds")
counts <- MRcounts(MRobj,norm=FALSE,log=FALSE)
nasal_clust3 <- FitGoM(t(counts), K=c(3), tol=0.00001, num_trials = 100)
nasal_clust4 <- FitGoM(t(counts), K=c(4), tol=0.00001, num_trials = 100)
nasal_clust5 <- FitGoM(t(counts), K=c(5), tol=0.00001, num_trials = 100)


fits <- list(counts_clust$clust_2, nasal_clust3$fit, nasal_clust4$fit, nasal_clust5$fit)
membership <- lapply(fits, function(xx) {
     apply(xx$omega, 1, which.max)
  })
names(membership) <- c("clust_2", "clust_3", "clust_4", "clust_5")

saveRDS(fits, file = "../output/hmp.Rmd/nasal_clust.rds")
saveRDS(membership, file = "../output/hmp.Rmd/nasal_clust_membership.rds")
```

```{r}
fits <- readRDS("../output/hmp.Rmd/nasal_clust.rds")
membership <- readRDS("../output/hmp.Rmd/nasal_clust_membership.rds")
table(membership$clust_2)
table(membership$clust_3)
table(membership$clust_4)
table(membership$clust_5)


pdf("../output/hmp.Rmd/nasal_clusts_structure.pdf")
for (i in 1:length(fits)) {
  annotation <- data.frame(
  sample_id = paste0("X", c(1:NROW(fits[[i]]$omega))),
  tissue_label = factor(membership[[i]],
                        levels = as.character(1:(i+1)) ) ) 
  rownames(fits[[i]]$omega) <- annotation$sample_id
  print(StructureGGplot(omega = fits[[i]]$omega,
                  annotation = annotation,
                  palette = RColorBrewer::brewer.pal(i+1, "Accent"),
                  yaxis_label = "Group",
                  order_sample = TRUE,
                  axis_tick = list(axis_ticks_length = .1,
                                   axis_ticks_lwd_y = .1,
                                   axis_ticks_lwd_x = .1,
                                   axis_label_size = 7,
                                   axis_label_face = "bold")) )
}
dev.off()
```

nasal analysis

```{r, eval=F}
MRobj = readRDS("../data/nasal_filtered_normed_batchcorrected.rds")
membership <- readRDS("../output/hmp.Rmd/nasal_clust_membership.rds")

pdf("../output/hmp.Rmd/nasal_clusts_taxa.pdf", width=10)

# get factor levels
pp <- plotBar.aggregate(MRobj[,order(colnames(MRobj))],lvl='Genus',n=20,
                  plot.factor_levels = levels(pp$dd$variable))
labs <- c(levels(pp$dd$variable)[-which(levels(pp$dd$variable)=="Other")], "Other")

getPalette = colorRampPalette(brewer.pal(9, "Set1"))
cols <- getPalette(21)
plotBar.aggregate(MRobj[,order(colnames(MRobj))],lvl='Genus',n=20,
                  plot.factor_levels = labs)$p +
  scale_fill_manual(values = cols) + xlab('Samples') +
#  theme(legend.text=element_text(size=7)) +
  guides(fill=guide_legend(ncol=1))


getPalette = colorRampPalette(brewer.pal(9, "Set1"))
plotBar(MRobj,lvl='Genus',cl=membership$clust_2,n=20)$p +
  scale_fill_manual(values = getPalette(21)) +
  guides(fill=guide_legend(ncol=1))

  
getPalette = colorRampPalette(brewer.pal(9, "Set1"))
plotBar(MRobj,lvl='Genus',cl=membership$clust_3,n=20)$p +
  scale_fill_manual(values = getPalette(21)) +
  guides(fill=guide_legend(ncol=1))
  

getPalette = colorRampPalette(brewer.pal(9, "Set1"))
plotBar(MRobj,lvl='Genus',cl=membership$clust_4,n=20)$p +
  scale_fill_manual(values = getPalette(21)) +
  guides(fill=guide_legend(ncol=1))
  
getPalette = colorRampPalette(brewer.pal(9, "Set1"))
plotBar(MRobj,lvl='Genus',cl=membership$clust_5,n=20)$p +
  scale_fill_manual(values = getPalette(21)) +
  guides(fill=guide_legend(ncol=1))

dev.off()
```



---


## Session information

```{r}
sessionInfo()
```

