---
title: "Nasal GOM"
author: "JNP"
output: html_document
---

<!-- The file analysis/chunks.R contains chunks that define default settings
shared across the workflowr files. -->
```{r read-chunk, include=FALSE, cache=FALSE}
knitr::read_chunk("chunks.R")
```

<!-- Update knitr chunk options -->
```{r knitr-opts-chunk, include=FALSE}
```

<!-- Insert the date the file was last updated -->
```{r last-updated, echo=FALSE, results='asis'}
```

<!-- Insert the code version (Git commit SHA1) if Git repository exists and R
 package git2r is installed -->
```{r code-version, echo=FALSE, results='asis'}
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning=F, message=F)
```

---

## Summary

$~$

---

## Package

```{r setup, include=FALSE}
library(knitr)
library(kableExtra)
library(dplyr)
library(metagenomeSeq))
library(CountClust)
```

## Loading data

Read in filtered data.

```{r}
MRobj = readRDS("../data/nasal_filtered.rds")
MRobj
```

$~$

---

## Normalization

We calculate the normalization scaling factors using cumulative sum scaling, see [Paulson et al](https://www.nature.com/articles/nmeth.2658) for more information.

```{r}
MRobj = readRDS("./nasal_filtered.rds")
MRobj = cumNorm(MRobj)
```

# Performing batch correction

Next, we perform species level batch correction on the raw counts. We then perform normalization on the batch corrected reads. We will update this section with an overview of the batch correction and reasoning.

```{r}
obj <- aggTax(MRobj,lvl="Species")
obj@expSummary$expSummary$normFactors <- normFactors(MRobj)

counts <- t(MRcounts(obj,norm=FALSE,log=FALSE))
batchcorrect_counts <- BatchCorrectedCounts(counts,
                                         batch_lab = pData(MRobj)$seqrun,
                                         use_parallel=FALSE)
counts = t(batchcorrect_counts)
```

# Clustering 

Next we perform clustering

```{r}
# nmat or counts?
counts_clust <- FitGoM(t(counts), K=c(2), tol=0.00001)

cct <- counts_clust[[1]]$omega
rowMaxs = sapply(seq(nrow(cct)),function(i){
    seq(along=cct[i,])[cct[i,]==max(cct[i,])]
  })
rowMaxs = unlist(rowMaxs)
uniqueGroups = unique(unlist(rowMaxs))
groupIds = split(seq(rowMaxs),factor(rowMaxs))

cl = rep("1",sum(sapply(groupIds,length))); cl[groupIds[[2]]] = 2

cbind(groupIDs = cl,round(cct,3)) %>% kable %>%
  kable_styling()
```

# Preserve data with GOM

Next we save the dataset with the grades of membership.

```{r}
MRobj2 = MRobj
assayData(MRobj2)[["counts"]] = counts
featureData(MRobj2) = AnnotatedDataFrame(data.frame(Species=rownames(counts)))
pData(MRobj)$GOM = cl
pData(MRobj2)$GOM = cl
save(MRobj2,MRobj,file="./nasal_filtered_GOM.rdata")
```

# Session Info

```{r}
sessionInfo()
```
